massa-execution/src/types.rs

```diff
@@ -26,6 +31,217 @@ pub(crate) struct StepHistoryItem {
 
     // list of SCE ledger changes caused by this execution step
     pub ledger_changes: SCELedgerChanges,
+
+    /// events produced during this step
+    pub events: EventStore,
+}
+
+/// Keep all events you need with some useful indexes
+#[derive(Default, Debug, Clone)]
+pub(crate) struct EventStore {
+    /// maps ids to events
+    id_to_event: Map<SCOutputEventId, SCOutputEvent>,
+
+    /// maps slot to a set of event ids
+    slot_to_id: HashMap<Slot, Set<SCOutputEventId>>,
+
+    /// maps initial caller to a set of event ids
+    caller_to_id: Map<Address, Set<SCOutputEventId>>,
+
+    /// maps direct event producer to a set of event ids
+    smart_contract_to_id: Map<Address, Set<SCOutputEventId>>,
+}
+
+impl EventStore {
+    /// add event to the store and all its indexes
+    pub fn insert(&mut self, id: SCOutputEventId, event: SCOutputEvent) {
+        if let Entry::Vacant(entry) = self.id_to_event.entry(id) {
+            self.slot_to_id
+                .get_mut(&event.context.slot)
+                .unwrap_or(&mut Set::<SCOutputEventId>::default())
+                .insert(id);
+            if let Some(caller) = event.context.call_stack.front() {
+                self.caller_to_id
+                    .get_mut(caller)
+                    .unwrap_or(&mut Set::<SCOutputEventId>::default())
+                    .insert(id);
+            }
+            if let Some(sc) = event.context.call_stack.back() {
+                self.smart_contract_to_id
+                    .get_mut(sc)
+                    .unwrap_or(&mut Set::<SCOutputEventId>::default())
+                    .insert(id);
+            }
+            entry.insert(event);
+        } else {
+            // just warn or return error ?
+            warn!("execution event already exist {:?}", id)
+        }
+    }
+
+    /// get just the map if ids to events
+    pub fn export(&self) -> Map<SCOutputEventId, SCOutputEvent> {
+        self.id_to_event.clone()
+    }
+
+    /// Clears the map, removing all key-value pairs. Keeps the allocated memory for reuse.
+    pub fn clear(&mut self) {
+        self.id_to_event.clear();
+        self.slot_to_id.clear();
+        self.caller_to_id.clear();
+        self.smart_contract_to_id.clear();
+    }
+
+    /// Remove exess events considering a config defined max
+    pub fn prune(&mut self, max_final_events: usize) {
+        // todo make setting static
+        if self.id_to_event.len() > max_final_events {
+            let diff = self.id_to_event.len() - max_final_events;
```

Comment:
You could store `len` in a variable to avoid calling it twice.

Reply:
changed

Reply:
Done